src/lib/statistics/chunk_statistics/histograms/equal_height_histogram.cpp

```diff
@@ -0,0 +1,199 @@
+  DebugAssert(min <= maxs[0], "Must have maxs and distinct counts for each bin.");
+
+  for (auto bin_id = 0u; bin_id < maxs.size(); bin_id++) {
+    DebugAssert(distinct_counts[bin_id] > 0, "Cannot have bins with no distinct values.");
+
+    if (bin_id < maxs.size() - 1) {
+      DebugAssert(maxs[bin_id] < maxs[bin_id + 1], "Bins must be sorted and cannot overlap.");
+    }
+  }
+}
+
+template <>
+EqualHeightHistogram<std::string>::EqualHeightHistogram(const std::vector<std::string>& maxs,
+                                                        const std::vector<HistogramCountType>& distinct_counts,
+                                                        const std::string& min, const HistogramCountType total_count,
+                                                        const std::string& supported_characters,
+                                                        const uint32_t string_prefix_length)
+    : AbstractHistogram<std::string>(supported_characters, string_prefix_length),
+      _maxs(maxs),
+      _distinct_counts(distinct_counts),
+      _min(min),
+      _total_count(total_count) {
+  DebugAssert(total_count > 0, "Cannot have histogram without any values.");
+  DebugAssert(maxs.size() > 0, "Cannot have histogram without any bins.");
+  DebugAssert(maxs.size() == distinct_counts.size(), "Must have maxs and distinct counts for each bin.");
+  DebugAssert(min <= maxs[0], "Must have maxs and distinct counts for each bin.");
+
+  for (auto bin_id = 0u; bin_id < maxs.size(); bin_id++) {
+    DebugAssert(distinct_counts[bin_id] > 0, "Cannot have bins with no distinct values.");
+    DebugAssert(maxs[bin_id].find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+
+    if (bin_id < maxs.size() - 1) {
+      DebugAssert(maxs[bin_id] < maxs[bin_id + 1], "Bins must be sorted and cannot overlap.");
+    }
+  }
+}
+
+template <typename T>
+EqualHeightBinStats<T> EqualHeightHistogram<T>::_get_bin_stats(
+    const std::vector<std::pair<T, HistogramCountType>>& value_counts, const BinID max_bin_count) {
+  const auto min = value_counts.front().first;
+  // If there are fewer distinct values than the number of desired bins use that instead.
+  const auto bin_count = max_bin_count <= value_counts.size() ? max_bin_count : value_counts.size();
+
+  // Bins shall have (approximately) the same height.
+  const auto total_count =
+      std::accumulate(value_counts.cbegin(), value_counts.cend(), HistogramCountType{0},
+                      [](HistogramCountType a, const std::pair<T, HistogramCountType>& b) { return a + b.second; });
+
+  // Make sure that we never create more bins than requested.
+  const auto count_per_bin = total_count / bin_count + (total_count % bin_count > 0u ? 1 : 0);
+
+  std::vector<T> maxs;
+  std::vector<HistogramCountType> distinct_counts;
+  maxs.reserve(bin_count);
+  distinct_counts.reserve(bin_count);
+
+  auto current_begin = 0ul;
```

Comment:
`= size_t{0}`, because these are indices into a vector. Also `current_begin` is too generic for me - is `current_bin_first_value_idx` (and similar for current_end and current_height) too long for you?